samples: mesh: nrf52: to solve pending issues

vikrant8052 opened the pull request:
1) Added Generic Default Transition Time Server & Client
Models names in list of Models supported by this App in README.rst

2) Initialized dev_uuid defined in ble_mesh.c to some arbitrary value.

3) Corrected conditional compilation directives defined in
publisher.c

4) Now it is possible to get proper target value & remaining
time in status response from Servers when transition in progress.
If transition is not in progress then those things would not
get integrate into status response. This is as per Bluetooth
Mesh Model specification. It is inspired concept from Bluetooth Mesh 
Developer study Guide.


Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8052 committed:
samples: mesh: nrf52: upgrade README.rst

Added Generic Default Transition Time Server & Client
Models names in list of Models supported by this App.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
@jhedberg 
Kindly do not merged this. I'm waiting for PTS logs from @MariuszSkamra to solve the pending issues.

jhedberg commented:
@kolodgrz once you're freed up from current PTS testing, could you run the relevant mesh model tests on the latest master branch for samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/ and report the findings to this PR or a separate issue. @vikrant8051 will then fix them. @vikrant8051 , @kolodgrz is handling PTS testing now for awhile since @MariuszSkamra is away.

dbkinder reviewed (approved):
doc change looks OK, but the commit message should be more meaningful

vikrant8052 committed:
samples: mesh: nrf52: initialized dev_uuid to some arbitrary value

Initialized dev_uuid defined in ble_mesh.c to some arbitrary value.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/ble_mesh.c:
```diff
@@ -33,7 +33,7 @@ static void prov_reset(void)
 }
 
-static u8_t dev_uuid[16];
+static u8_t dev_uuid[16] = { 0xdd, 0xdd };
```
Is this in case bt_le_oob_get_local() fails, so you don't end-up with an all-zeroes UUID in that case? If so, please state it in the commit message. (I wouldn't call it "random" btw. Arbitrary, perhaps).

vikrant8051 replied:
bt_le_oob_get_local() is working properly after yesterday's patch from your side.
I just set dev_uuid[16] to { 0xdd, 0xdd } to mimic onoff-app.

kolodgrz commented:
@jhedberg I didn't reviewed this app before and don't have test plan for this. So first of all I have to create test plan. Is it somewhere documented, to not double work ?

jhedberg commented:
> @jhedberg I didn't reviewed this app before and don't have test plan for this. So first of all I have to create test plan. Is it somewhere documented, to not double work ?

@MariuszSkamra already did a few runs of these tests, but I'm not sure of where his test plan is or if he created any documentation for it.

vikrant8051 commented:
@kolodgrz Please wait & don't yet execute PTS. I have to change things as per latest feedback received from PTS member from SIG. I will inform you once I do with my changes. Thank You !!

vikrant8052 committed:
samples: mesh: nrf52: revised status resp. behaviour & TT calculation

With this it is possible to get proper target value & remaining
time in status response from Servers when transition in progress.
If transition is not in progress then those things would not
get integrate into status response. This is as per Bluetooth
Mesh Model specification.

It is inspired concept from Bluetooth Mesh Developer study
Guide.

Also updated Transition Time (TT) related parameter calculation.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8052 committed:
samples: mesh: nrf52: corrected conditional compilation

Corrected conditional compilation directives defined in
publisher.c

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/transition.c:
```diff
@@ -507,6 +602,11 @@ K_WORK_DEFINE(light_lightness_linear_work, light_lightness_linear_work_handler);
+void rt_handler(struct k_timer *dummy)
+{
+	rt_counter--;
+}
```
Wouldn't it be more power efficient to just store the "start" as a time-stamp and then do the subtraction based on the elapsed time when it's actually needed, instead of having this timer firing off every 100ms?

vikrant8051 replied:
Timer of 100ms will turn off with Transition End.

vikrant8051 replied:
Timer overflow is going to be issue ....if we consider the way you have suggested

vikrant8051 replied:
Please wait ...don't review ...since I'm correcting the things.

vikrant8051 replied:
I'm trying to implement as per your suggetion

vikrant8051 replied:
Done !!
I've implemented as per your suggestion. Yes it is accurate or correct implementation than mine !!

vikrant8051 commented:
@jhedberg Things are ready. Please review now.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/transition.c:
```diff
@@ -14,8 +14,46 @@
 
-u8_t enable_transition;
-u8_t default_tt;
+static bool is_new_transition_start;
```
Isn't the transition bound to a specific state, i.e. instead of being global shouldn't this variable be part of the state struct(s)?

vikrant8051 replied:
Currently, only one transition is allowed at a time. Using global variable here reduces complexity.
Here I wanna wait & see what PTS is expecting.

vikrant8051 replied:
Ok ... I will put "is_new_transition_start" in each relevant struct.

vikrant8051 replied:
Done !!

jhedberg replied:
I don't see this change. The start timestamp is still external to any structs.

vikrant8051 replied:
"is_new_transition_start" I was assuming about this variable. I've created separate entry in Stuct define for Servers

But now not getting what your are expecting ? Could you elaborate more on it or directly mentioned about that variable ?

vikrant8051 replied:
Done !!

vikrant8052 committed:
samples: mesh: nrf52: corrected spelling

Corrected spelling mistakes for content of enum defined in
transition.h & renamed some functions. Corrected comments
in main.c

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

jhedberg commented:
@sjanc @kolodgrz what's the status on running the PTS tests on this?

vikrant8051 commented:
@carlescufi @jhedberg 
After syncing with Main branch, during testing I got this fault: 

***** MPU FAULT *****                                                                                                             
  Instruction Access Violation                                                                                                    
***** Hardware exception *****                                                                                                    
Current thread ID = 0x20001884                                                                                                    
Faulting instruction address = 0x20001c70                                                                                         
Fatal fault in ISR! Spinning...


I'm using buttons to publish messages to Servers running on same board (nrf52_PDK). Before latest sync, everything was working perfectly.

I executed following command for debugging -

**/opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/bin/arm-zephyr-eabi/arm-zephyr-eabi-addr2line -e /home/vikrant/zephyr/samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/build/zephyr/zephyr.elf 0x20001c70**

o/p -> :?

vikrant8051 commented:
@carlescufi @jhedberg 
I don't know how, but previously mentioned MPU fault issue has solved !!

vikrant8051 commented:
Hi, Any update regarding PTS ?
Thank You  !!

vikrant8052 committed:
samples: mesh: nrf52: corrected state Binding for Delta Level

Now if transition (instantaneous or non-instantaneous) is due
to delta level, then state binding would be as per it.

This is as per 3.3.2.2.3 of Bluetooth Mesh Model
Specification.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
@carlescufi @jhedberg 
Once again got following fault

***** MPU FAULT *****
  Instruction Access Violation
***** Hardware exception *****
Current thread ID = 0x2000188c
Faulting instruction address = 0x20001c5c
Fatal fault in ISR! Spinning...

MariuszSkamra commented:
Enable stack analysis to check stacks usage. I had similar issue some time ago and the reason was too small system workqueue size.

vikrant8051 commented:
@MariuszSkamra 
Thanks for solution. Let me check it !!

vikrant8051 commented:
@MariuszSkamra 
After setting CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=4096, issue has solved.
Thanks for the solution.

vikrant8051 commented:
@MariuszSkamra 
It come back !! With actual lights (with PWM) getting MPU fault if received message contains transition time & delay fields.
It was not under my observation before syncing with master branch.

MariuszSkamra commented:
List of FAILING test cases:
MMDL/SR/GOO/BV-02-C https://hastebin.com/ahaderunij.cs https://hastebin.com/ositixurik.sql
MMDL/SR/GLV/BV-02-C https://hastebin.com/tatoqimuxo.shell https://hastebin.com/hunozuyuzu.sql
MMDL/SR/GLV/BV-03-C https://hastebin.com/iwutunozuh.shell https://hastebin.com/wofiwucumi.sql
MMDL/SR/GLV/BV-05-C https://hastebin.com/ikevitiduh.cs https://hastebin.com/ezaberujev.sql
MMDL/SR/GLV/BV-06-C https://hastebin.com/ramasiboya.cs https://hastebin.com/rejapigona.php
MMDL/SR/LLN/BV-06-C https://hastebin.com/imizuxepuz.pl https://hastebin.com/mediwoduga.sql
MMDL/SR/LLN/BV-07-C https://hastebin.com/yepigoyuta.pl https://hastebin.com/ajukokoqin.sql
MMDL/SR/LLN/BV-08-C https://hastebin.com/zilosayibe.pl https://hastebin.com/ocayekecul.sql
MMDL/SR/LLN/BV-09-C https://hastebin.com/ayubofabom.pl https://hastebin.com/nekedobewu.sql
MMDL/SR/LLN/BV-10-C https://hastebin.com/fequwelinu.pl https://hastebin.com/afomogiqox.sql
MMDL/SR/LLN/BV-11-C https://hastebin.com/teraratuda.pl https://hastebin.com/uqewahunuk.sql
MMDL/SR/LLN/BV-12-C https://hastebin.com/ujonecefed.pl https://hastebin.com/adexujixak.sql
MMDL/SR/LLN/BV-13-C https://hastebin.com/ataxujakan.pl https://hastebin.com/nuxapejuqa.sql
MMDL/SR/LLN/BV-15-C https://hastebin.com/renomoyuho.shell https://hastebin.com/sajutajihe.sql
MMDL/SR/LLN/BV-16-C https://hastebin.com/okovagohum.shell https://hastebin.com/tajixacuko.sql
MMDL/SR/LLN/BV-17-C https://hastebin.com/jocafararo.pl https://hastebin.com/ehujafequk.sql
MMDL/SR/LCTL/BV-04-C https://hastebin.com/ququdetomo.pl https://hastebin.com/ufagubazim.sql
MMDL/SR/LCTL/BV-05-C https://hastebin.com/afebagasij.shell https://hastebin.com/ibecometax.sql
MMDL/SR/LCTL/BV-06-C https://hastebin.com/onaciyolaz.pl https://hastebin.com/rifamecayi.sql
MMDL/SR/LLNS/BI-01-C https://hastebin.com/ifunopegex.pl https://hastebin.com/zufosejuzo.sql
MMDL/SR/LCTLS/BI-01-C https://hastebin.com/icedihiyes.pl https://hastebin.com/aqobolezof.sql

vikrant8051 commented:
@MariuszSkamra Thanks for latest logs.

vikrant8051 commented:
- [ ] MMDL/SR/GOO/BV-02-C https://hastebin.com/ahaderunij.cs https://hastebin.com/ositixurik.sql
- [ ] MMDL/SR/GLV/BV-02-C https://hastebin.com/tatoqimuxo.shell https://hastebin.com/hunozuyuzu.sql
- [ ] MMDL/SR/GLV/BV-03-C https://hastebin.com/iwutunozuh.shell https://hastebin.com/wofiwucumi.sql
- [ ] MMDL/SR/GLV/BV-05-C https://hastebin.com/ikevitiduh.cs https://hastebin.com/ezaberujev.sql
- [ ] MMDL/SR/GLV/BV-06-C https://hastebin.com/ramasiboya.cs https://hastebin.com/rejapigona.php
- [ ] MMDL/SR/LLN/BV-06-C https://hastebin.com/imizuxepuz.pl https://hastebin.com/mediwoduga.sql
- [ ] MMDL/SR/LLN/BV-07-C https://hastebin.com/yepigoyuta.pl https://hastebin.com/ajukokoqin.sql
- [ ] MMDL/SR/LLN/BV-08-C https://hastebin.com/zilosayibe.pl https://hastebin.com/ocayekecul.sql
- [ ] MMDL/SR/LLN/BV-09-C https://hastebin.com/ayubofabom.pl https://hastebin.com/nekedobewu.sql
- [ ] MMDL/SR/LLN/BV-10-C https://hastebin.com/fequwelinu.pl https://hastebin.com/afomogiqox.sql
- [ ] MMDL/SR/LLN/BV-11-C https://hastebin.com/teraratuda.pl https://hastebin.com/uqewahunuk.sql
- [ ] MMDL/SR/LLN/BV-12-C https://hastebin.com/ujonecefed.pl https://hastebin.com/adexujixak.sql
- [ ] MMDL/SR/LLN/BV-13-C https://hastebin.com/ataxujakan.pl https://hastebin.com/nuxapejuqa.sql
- [ ] MMDL/SR/LLN/BV-15-C https://hastebin.com/renomoyuho.shell https://hastebin.com/sajutajihe.sql
- [ ] MMDL/SR/LLN/BV-16-C https://hastebin.com/okovagohum.shell https://hastebin.com/tajixacuko.sql
- [ ] MMDL/SR/LLN/BV-17-C https://hastebin.com/jocafararo.pl https://hastebin.com/ehujafequk.sql
- [ ] MMDL/SR/LCTL/BV-04-C https://hastebin.com/ququdetomo.pl https://hastebin.com/ufagubazim.sql
- [ ] MMDL/SR/LCTL/BV-05-C https://hastebin.com/afebagasij.shell https://hastebin.com/ibecometax.sql
- [ ] MMDL/SR/LCTL/BV-06-C https://hastebin.com/onaciyolaz.pl https://hastebin.com/rifamecayi.sql
- [ ] MMDL/SR/LLNS/BI-01-C https://hastebin.com/ifunopegex.pl https://hastebin.com/zufosejuzo.sql
- [ ] MMDL/SR/LCTLS/BI-01-C https://hastebin.com/icedihiyes.pl https://hastebin.com/aqobolezof.sql

vikrant8051 commented:
@jhedberg @MariuszSkamra 
I've corrected issue with "marked" logs. For that I've moved
state->transition_start_timestamp = k_uptime_get(); from Server's message handler to Server's timer thread. Please review recent changes.
--------------------------------------------------------------------------------------------------------------------------------
MMDL/SR/LLN/BV-06-C https://hastebin.com/mediwoduga.sql
MMDL/SR/LLN/BV-07-C https://hastebin.com/ajukokoqin.sql
MMDL/SR/LLN/BV-08-C https://hastebin.com/ocayekecul.sql
MMDL/SR/LLN/BV-09-C https://hastebin.com/nekedobewu.sql
MMDL/SR/LLN/BV-10-C https://hastebin.com/afomogiqox.sql
MMDL/SR/LLN/BV-11-C https://hastebin.com/uqewahunuk.sql
MMDL/SR/LLN/BV-12-C https://hastebin.com/adexujixak.sql
MMDL/SR/LLN/BV-13-C https://hastebin.com/nuxapejuqa.sql
MMDL/SR/LCTL/BV-04-C https://hastebin.com/ufagubazim.sql

I think these are exaggerated expectation becase ..
in case of **MMDL/SR/LLN/BV-06-C**...demands Light_Lightness_Linear status response &
in case of **MMDL/SR/LLN/BV-07-C**...demands Generic Level status response
after sending "Light Lightness Set".

I don't where this is explained in Mesh Model Specification.

vikrant8052 committed:
samples: mesh: nrf52: corrected Range status response

As per PTS, IUT is expected to return success for the
Range Get messages (Light Lightness Range Get for
MMDL/SR/LLNS/BI-01-C, Light CTL Temperature Range Get
for MMDL/SR/LCTLS/BI-01-C). This commit take care of
this issue.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
@jhedberg
If possible as of now merge this PR after review. Now if execute PTS again, then it might return 
9 issues which are regarding **extra** status responses + 2 which is due to unable to save Servers state on SoC flash.

jhedberg commented:
@vikrant8051 yep, this PR has gone on for too long IMO. CI passes now, so I'll merge it. We can then sort out remaining issues in a new PR.

vikrant8051 commented:
@jhedberg Thanks for merging it.
It will be great if we once again execute PTS against this App.
If we get only previously mentioned 9+2 issues, that will confirm current architecture is stable.
